cmake/EnableCompilerWarnings.cmake

```diff
@@ -26,89 +26,104 @@
 # - Enable warning all for gcc/clang or use /W4 for visual studio
 
 ## Strict warning level
-if (MSVC)
-    # Use the highest warning level for visual studio.
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /w")
-    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /w")
-    # set(CMAKE_CXX_WARNING_LEVEL 4)
-    # if (CMAKE_CXX_FLAGS MATCHES "/W[0-4]")
-    #     string(REGEX REPLACE "/W[0-4]" "/W4" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
-    # else ()
-    #     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /W4")
-    # endif ()
+set(__msvc_cxx_compile_options /W4)
 
-    # set(CMAKE_C_WARNING_LEVEL 4)
-    # if (CMAKE_C_FLAGS MATCHES "/W[0-4]")
-    #     string(REGEX REPLACE "/W[0-4]" "/W4" CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")
-    # else ()
-    #     set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /W4")
-    # endif ()
+set(__default_cxx_compile_options
+    -Wall
+    -Wextra
+    -Wcomment
+    -Wendif-labels
+    -Wformat
+    -Winit-self
+    -Wreturn-type
+    -Wsequence-point
+    -Wswitch
+    -Wtrigraphs
+    -Wno-undef
+    -Wuninitialized
+    -Wunreachable-code
+    -Wunused
+    -Wno-ignored-qualifiers
+    -Wno-sign-compare
+)
 
-else()
-    foreach(COMPILER C CXX)
-        set(CMAKE_COMPILER_WARNINGS)
-        # use -Wall for gcc and clang
-        list(APPEND CMAKE_COMPILER_WARNINGS
-            -Wall
-            -Wextra
-            -Wcomment
-            -Wendif-labels
-            -Wformat
-            -Winit-self
-            -Wreturn-type
-            -Wsequence-point
-            # Shadow is broken on gcc when using lambdas
-            # -Wshadow
-            -Wswitch
-            -Wtrigraphs
-            -Wundef
```

Comment:
This must be reverted. We do not allow usage of undefined macros.

Reply:
@apwojcik could you follow up with this one? Currently, we only disable it for third party like `GTest`

Reply:
done